packages/backend-storage/src/storage_access_arbiter.ts

```diff
@@ -0,0 +1,145 @@
+        const allowSet = new Set<StoragePath>([s3Prefix]);
+        const denySet = new Set<StoragePath>();
+        accessMap.set(action, { allow: allowSet, deny: denySet });
+
+        // this is where we create the reverse mapping that allows us to add entries to the denySet later by looking up the prefix
+        this.setPrefixDenyMapEntry(s3Prefix, allowSet, denySet);
+      } else {
+        // otherwise add the prefix to the existing allow set
+        accessMap.get(action)?.allow.add(s3Prefix);
+      }
+    });
+  };
+
+  /**
+   * Iterates over all of the access definitions that have been added to the translator,
+   * generates a policy for each accessMap,
+   * and attaches the policy to the corresponding ResourceAccessAcceptor
+   *
+   * After this method is called, the existing access definition state is cleared.
+   * This prevents multiple calls to this method from producing duplicate policies.
+   * The class can continue to be used to build up state for a new set of policies if desired.
+   * @param ssmEnvironmentEntries Additional SSM context that is passed to each ResourceAccessAcceptor
+   */
+  attachPolicies = (ssmEnvironmentEntries: SsmEnvironmentEntry[]) => {
+    const allPaths = Array.from(this.prefixDenyMap.keys());
+    allPaths.forEach((storagePath) => {
+      const parent = findParent(storagePath, allPaths);
+      if (!parent) {
+        return;
+      }
+      // if a parent path is defined, invoke the denyByDefault callback on this subpath for all policies that exist on the parent path
+      this.prefixDenyMap
+        .get(parent)
+        ?.forEach((denyByDefaultCallback) =>
+          denyByDefaultCallback(storagePath)
+        );
+    });
+
+    this.acceptorAccessMap.forEach(({ acceptor, accessMap }) => {
+      acceptor.acceptResourceAccess(
+        this.policyFactory.createPolicy(accessMap),
+        ssmEnvironmentEntries
+      );
+    });
+    this.acceptorAccessMap.clear();
+    this.prefixDenyMap.clear();
+  };
+
+  private setPrefixDenyMapEntry = (
+    storagePath: StoragePath,
+    allowPathSet: Set<StoragePath>,
+    denyPathSet: Set<StoragePath>
+  ) => {
+    // function that will add the denyPath to the denyPathSet unless the allowPathSet explicitly allows the path
+    const setDenyByDefault = (denyPath: StoragePath) => {
+      if (!allowPathSet.has(denyPath)) {
+        denyPathSet.add(denyPath);
+      }
+    };
+    if (!this.prefixDenyMap.has(storagePath)) {
+      this.prefixDenyMap.set(storagePath, [setDenyByDefault]);
+    } else {
+      this.prefixDenyMap.get(storagePath)?.push(setDenyByDefault);
+    }
+  };
+}
+
+/**
+ * Returns the element in paths that is a prefix of path, if any
+ * Note that there can only be one at this point because of upstream validation
+ */
+const findParent = (path: string, paths: string[]) =>
+  paths.find((p) => path !== p && path.startsWith(p.replaceAll('*', ''))) as
```

Comment:
two questions

do we assure somewhere that `*` is always a suffix?
I.e. is `/foo/*/bar/*` possible ?

Is `/foo/bar*` (i.e. no slash before `*`) handled properly (is this even a valid case?) ?

Reply:
Both of those are invalid paths based on our current validator: https://github.com/aws-amplify/amplify-backend/blob/215d65d4952217431b913ac456217851dbd21524/packages/backend-storage/src/validate_storage_access_paths.ts#L19-L26

However, I am thinking that it might make sense to move the validation logic inside of this class so we aren't relying on some other upstream class to run validation that ensures correct behavior of this class.

Reply:
or perhaps extract `class StoragePathValidator` and apply in both places.

Reply:
I moved all of the "arbiter" class inside of the "orchestrator". The logic is much more colocated now and I think it reads a lot better. [This method](https://github.com/aws-amplify/amplify-backend/pull/1070/files#diff-59ceb8af2428bc2169787051bf9f2123929806c2a0e7687cb414d60dfdd5643fR68) provides a single entry point for all of the access rule evaluation and doesn't depend on any other method "doing the right thing"